armory/instrument/instrument.py

```diff
@@ -0,0 +1,693 @@
+        self.final_kwargs = final_kwargs or {}
+        if not isinstance(self.final_kwargs, dict):
+            raise ValueError(f"final_kwargs must be None or a dict, not {final_kwargs}")
+        self.record_final_only = record_final_only
+        self.never_measured = True
+
+    def get_arg_names(self):
+        return list(self.arg_index)
+
+    def clear(self):
+        """
+        Clear all of the current values
+        """
+        self.values = [None] * len(self.arg_index)
+        self.values_set = [False] * len(self.arg_index)
+        self.arg_batch_indices = [None] * len(self.arg_index)
+
+    def add_writer(self, writer):
+        """
+        Add a writer to emit outputs to.
+        """
+        if writer not in self.writers:
+            self.writers.append(writer)
+
+    def set(self, name, value, batch):
+        """
+        Set the value for the arg name
+        """
+        if name not in self.arg_index:
+            raise ValueError(f"{name} not a valid arg name from {self.arg_index}")
+        i = self.arg_index[name]
+        self.values[i] = value
+        self.values_set[i] = True
+        self.arg_batch_indices[i] = batch
+        if self.auto_measure and self.is_ready():
+            self.measure()
+
+    def is_ready(self, raise_error=False):
+        """
+        Return True if all values have been set and batch numbers match
+            if raise_error is True, raise ValueError instead of returning False
+        """
+        if not all(self.values_set):
+            if raise_error:
+                raise ValueError(f"Not all values have been set: {self.values_set}")
+            return False
+        if any(self.arg_batch_indices[0] != batch for batch in self.arg_batch_indices):
+            if raise_error:
+                raise ValueError(
+                    "Batch numbers are mismatched: {self.arg_batch_indices}"
+                )
+            return False
+        return True
+
+    def measure(self, clear_values=True):
+        self.is_ready(raise_error=True)
+        result = self.metric(*self.values, **self.metric_kwargs)
+        record = (self.name, self.arg_batch_indices[0], result)
+        # Assume metric is sample-wise, but computed on a batch of samples
+        try:
+            self._results.extend(result)
+        except TypeError:
+            self._results.append(result)
+        if not self.record_final_only:
+            for writer in self.writers:
+                writer.write(record)
+        if clear_values:
+            self.clear()
+        self.never_measured = False
+
+    def finalize(self):
+        """
+        Primarily intended for metrics of metrics, like mean or stdev of results
+        """
+        if self.never_measured:
+            unset = [arg for arg, i in self.arg_index.items() if not self.values_set[i]]
+            log.warning(
+                f"Meter '{self.name}' was never measured. "
+                f"The following args were never set: {unset}"
```

Comment:
so we have a warning if a meter is never measured, but would it also make sense to have a warning if a probe value "falls on the floor" i.e. is never used to measure anything / isn't connected to a meter?

Reply:
If no sink is connected to the probe, then it logs a warning. I can think of lots of valid reasons a probe value would not be measured, however, due to filtering by the meters.